libs/formats/src/zone/zone-validator.ts

```diff
@@ -0,0 +1,325 @@
+
+  if (!volumes) throw new ZoneValidationError('Missing volumes');
+
+  if (formatVersion !== CURRENT_ZONE_FORMAT_VERSION)
+    throw new ZoneValidationError('Bad format version');
+
+  validateVolumes(volumes);
+
+  const usedVolumes = Array.from({ length: volumes.length }).fill(
+    false
+  ) as boolean[];
+
+  let totalZones = 0;
+
+  validateTrack(tracks.main, 'Main', volumes);
+
+  if (tracks.main.zones.segments.length === 1) {
+    if (tracks.stages.length > 0)
+      throw new ZoneValidationError(
+        'A map with a linear main track cannot have any stage stacks'
+      );
+  } else {
+    if (tracks.stages.length !== tracks.main.zones.segments.length) {
+      throw new ZoneValidationError(
+        'The number of stage tracks must match the number of main track segments'
+      );
+    }
+  }
+
+  if (tracks.stages.length > 0) {
+    if (tracks.stages.length > MAX_STAGE_TRACKS)
+      throw new ZoneValidationError('Too many stage tracks');
+
+    for (const [stageIndex, stage] of tracks.stages.entries()) {
+      if (stage?.zones?.segments?.length !== 1)
+        throw new ZoneValidationError(
+          `Stage ${stageIndex + 1} track must have a single segment`
+        );
+
+      validateTrack(stage, `stage ${stageIndex}`, volumes);
+    }
+  }
+
+  if (tracks.bonuses.length > 0) {
+    if (tracks.bonuses.length > MAX_BONUS_TRACKS)
+      throw new ZoneValidationError('Too many bonus tracks');
+
+    for (const [bonusIndex, bonus] of tracks.bonuses.entries()) {
+      if (bonus?.zones?.segments?.length !== 1)
+        throw new ZoneValidationError(
+          `Bonus ${bonusIndex + 1} track must have a single segment`
+        );
+
+      validateTrack(bonus, `bonus ${bonusIndex}`, volumes);
+    }
+  }
+
+  if (totalZones > MAX_ZONES_ALL_TRACKS)
+    throw new ZoneValidationError('Too many zones in total');
+
+  for (const [i, isUsed] of usedVolumes.entries()) {
+    if (!isUsed) throw new ZoneValidationError(`Volume ${i} is unused`);
+  }
+
+  function validateVolumes(volumes: Volume[]) {
+    for (const [volumeIndex, volume] of volumes.entries()) {
+      if (volume.regions.length === 0) {
+        throw new ZoneValidationError(`Volume ${volumeIndex} has no regions`);
+      }
+
+      validateRegions(volume.regions, volumeIndex);
+    }
+  }
+
+  function validateRegions(regions: Region[], volumeIndex: number) {
+    for (const [regionIndex, region] of regions.entries()) {
+      if (region.points.length < 3)
+        throw new ZoneValidationError(
+          `Volume ${volumeIndex} region ${regionIndex} does not have enough points`
+        );
+
+      if (region.points.length > MAX_ZONE_REGION_POINTS)
+        throw new ZoneValidationError(
+          `Volume ${volumeIndex} region ${regionIndex} has too many points`
+        );
+
+      // Using non-strict as JS coerces to value === undefined || value === null
+      if (region.bottom == null || region.bottom >= FLT_MAX) {
```

Comment:
We don't *really* need this `FLT_MAX` check, since it's just used in C++ to detect that the JSON field was null or non-existent. If the field is actually present and set to `FLT_MAX`, it would be caught by the out-of-bounds check below, which would make more sense anyway.